packages/tools/devtools/devtools-core/src/data-visualization/SharedTreeVisualizer.ts

```diff
@@ -0,0 +1,330 @@
+			children[key] = child;
+		}
+
+		return {
+			children,
+			nodeKind: VisualNodeKind.TreeNode,
+			tooltipContents: tree.schema.allowedTypes,
+		};
+	}
+}
+
+/**
+ * Returns the allowed fields & types for the object fields.
+ */
+function getObjectAllowedTypes(schema: ObjectNodeStoredSchema): string {
+	let result = "";
+
+	for (const [fieldKey, treeFieldStoredSchema] of schema.objectNodeFields) {
+		// Set of allowed tree types {@link TreeTypeSet}.
+		const fieldTypes = treeFieldStoredSchema.types;
+
+		let fieldAllowedType = fieldKey === "" ? "" : `${fieldKey} : `;
+
+		// If not specified, types are unconstrained.
+		if (fieldTypes === undefined) {
+			fieldAllowedType += "any";
+		} else {
+			for (const type of fieldTypes) {
+				fieldAllowedType += `${type} | `;
+			}
+		}
+
+		// Slice the trailing ` | ` from the `fieldAllowedType`.
+		fieldAllowedType = `${fieldAllowedType.slice(0, -3)}, `;
+
+		result += fieldAllowedType;
+	}
+
+	// Slice the trailing `, ` from the `result`.
+	result = result.slice(0, -2);
+
+	return `{ ${result} }`;
+}
+
+/**
+ * Returns the allowed fields & types for the map fields.
+ */
+function getMapAllowedTypes(
+	fields: FieldMapObject<JsonableTree> | undefined,
+	schema: MapNodeStoredSchema,
+): string {
+	let fieldAllowedTypes = "";
+
+	const mapFieldAllowedTypes = schema.mapFields.types;
+
+	if (mapFieldAllowedTypes === undefined) {
+		fieldAllowedTypes = "any";
+	} else {
+		for (const type of mapFieldAllowedTypes) {
+			fieldAllowedTypes += `${type} | `;
+		}
+	}
+
+	// Slice the trailing ` | ` from the `fieldAllowedType`.
+	fieldAllowedTypes = fieldAllowedTypes.slice(0, -3);
+
+	if (fields === undefined) {
+		throw new TypeError("Fields should not be undefined.");
+	}
+
+	let result = "";
+	for (const [fieldKey] of Object.entries(fields)) {
+		result += `${fieldKey} : ${fieldAllowedTypes}, `;
+	}
+
+	// Slice the trailing `, ` from the `result`.
+	result = result.slice(0, -2);
+
+	return `{ ${result} }`;
+}
+
+/**
+ * Returns the allowed fields & types for the leaf fields.
+ */
+function getLeafAllowedTypes(schema: TreeFieldStoredSchema): string {
+	let result = "";
+	const leafTypes = schema.types;
+
+	if (leafTypes === undefined) {
+		result = "any";
+	} else {
+		for (const type of leafTypes) {
+			result += `${type} | `;
+		}
+	}
+
+	// Slice the trailing ` | ` from the `result`.
+	result = `${result.slice(0, -3)}`;
+
+	return result;
+}
+
+/**
+ * Returns the schema & leaf value of the node with type {@link LeafNodeStoredSchema}.
+ */
+function visualizeLeafNodeStoredSchema(
+	tree: JsonableTree,
+	allowedTypes: string,
+): SharedTreeLeafNode {
+	return {
+		schema: {
+			schemaType: SharedTreeSchemaType.LeafNodeStoredSchema,
+			allowedTypes,
+		},
+		value: JSON.stringify(tree.value),
+		kind: VisualSharedTreeNodeKind.LeafNode,
+	};
+}
+
+/**
+ * Returns the schema & fields of the node with type {@link ObjectNodeStoredSchema}.
+ */
+function visualizeObjectNodeStoredSchema(
```

Comment:
Nit: since this function visualizes an object node (and not just its schema), I would recommend naming this something like `visualizeObjectNode`.